vspd.go

```diff
@@ -0,0 +1,543 @@
+package dcrlibwallet
+
+import (
+	"bytes"
+	"crypto/ed25519"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/asdine/storm"
+	"github.com/decred/dcrd/blockchain/stake/v3"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v2"
+	"github.com/decred/dcrd/dcrutil/v3"
+	"github.com/decred/dcrd/txscript/v2"
+	"github.com/decred/dcrd/wire"
+)
+
+type VSPD struct {
+	baseURL             string
+	pubKey              []byte
+	w                   *Wallet
+	mwRef               *MultiWallet
+	sourceAccountNumber int32
+	httpClient          *http.Client
+	params              *chaincfg.Params
+}
+
+func (mw *MultiWallet) NewVSPD(baseURL string, walletID int, sourceAccountNumber int32) *VSPD {
+	sourceWallet := mw.WalletWithID(walletID)
+	if sourceWallet == nil {
+		return nil
+	}
+
+	return &VSPD{
+		baseURL:             baseURL,
+		w:                   sourceWallet,
+		mwRef:               mw,
+		sourceAccountNumber: sourceAccountNumber,
+		httpClient:          new(http.Client),
+		params:              mw.chainParams,
+	}
+}
+
+// GetInfo returns the information of the specified VSP base URL
+func (v *VSPD) GetInfo() (*GetVspInfoResponse, error) {
+	resp, err := v.httpClient.Get(v.baseURL + "/api/v3/vspinfo")
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("Non 200 response from server: %v", string(b))
+	}
+
+	var vspInfoResponse GetVspInfoResponse
+	err = json.Unmarshal(b, &vspInfoResponse)
+	if err != nil {
+		return nil, err
+	}
+	v.pubKey = vspInfoResponse.PubKey
+
+	err = validateVSPServerSignature(resp, v.pubKey, b)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Thing is the validateVSPServerSignature() requires the pubkey as a parameter for validating the server signature. the function is called by all the vspd functions through the signedVSP_HTTP() hence, it cannot be set after the validation. The pubKey from the getinfo() is recommended to be cached so it can be passed to calling code on request as signature validation.

Reply:
It can be set after validation. I'm thinking of what bug could come in if the `v.pubKey` is set and validation fails.
I edited the suggestion code to use `vspInfoResponse.PubKey, `v.pubKey' is set right after validating so you would have no issue calling other functions that requires it to be set.